### PR TITLE
Make aten::as_strided() as a view.

### LINKED ISSUE #870: Loss doesn't backprop through copy-based reshapes
Simple repro: change this line to `x = torch.flatten(x, 1)`:
https://github.com/pytorch/xla/blob/master/test/test_train_mnist.py#L43

Then add this to train loop:
`print("conv1_max: %s  conv1_min: %s  fc1_max: %s  fc1_min: %s" % (str(model.conv1.weight.max().item()), str(model.conv1.weight.min().item()), str(model.fc1.weight.max().item()), str(model.fc1.weight.min().item())))`

The conv weights in that print statement do not change during training when using torch.flatten or torch.reshape, but the weights do change if using the original line: `x = x.view(-1, 320)`

view() returns a reference to the original tensor whereas flatten/reshape return a reference to a copy of the original tensor. Note from that print statement above that the fully-connected layer weights (which occur after the reshape in the forward pass) do change during training whereas the conv weights (which occur before the reshape in the forward pass) do not change

### PR DESCRIPTION
**Pull request by dlibenzi**
Fixes #870

### CODE REVIEW

File: test/cpp/test_aten_xla_tensor.cpp [hunk 1]
```diff
@@ -6571,6 +6569,20 @@ TEST_F(AtenXlaTensorTest, TestAsStridedWithOffset) {
...
 
+TEST_F(AtenXlaTensorTest, TestAsStridedWithInplaceCopy) {
+  torch::Tensor grad = torch::ones({4}, torch::TensorOptions(torch::kFloat));
+  std::vector<int64_t> size = {4};
+  std::vector<int64_t> stride = {1};
+  torch::Tensor output = torch::zeros({4}, grad.options());
+  output.as_strided(size, stride).copy_(grad);
+  ForEachDevice([&](const torch::Device& device) {
+    torch::Tensor xla_grad = CopyToDevice(grad, device);
+    torch::Tensor xla_output = torch::zeros({4}, xla_grad.options());
```
Comment (zcain117) [inline test/cpp/test_aten_xla_tensor.cpp]: This hasn't been copied to device, so I think the line below is testing:
<tensor not on device>.as_strided(size, stride).copy_(<tensor on device>);

We might want to test either instead or in addition:
<tensor on device>.as_strided(size, stride).copy_(<tensor on device>);

**Comment by asuhan** [comment]
Is the test failure legitimate?

File: test/cpp/test_aten_xla_tensor.cpp (hunk 1 above)
Comment (dlibenzi) [reply test/cpp/test_aten_xla_tensor.cpp]: xla_grad.options() contains the device, so the xla_output is on XLA device.
Otherwise the copy below would ask a pytorch CPU tensor to copy data from an XLA one, and it'd fail.

**Comment by dlibenzi** [comment]
> Is the test failure legitimate?

It was. I had a change sneaked in which I forgot to revert.

File: torch_xla/csrc/view.h [hunk 2]
```diff
@@ -22,6 +22,14 @@ struct SelectInfo {
   xla::int64 stride = 0;
 };
 
+struct AsStridedInfo {
```
Comment (asuhan) [inline torch_xla/csrc/view.h]: Do you like it better as a struct? It only has one field and we end up with a few lines of boilerplate because of it.

Comment (dlibenzi) [reply torch_xla/csrc/view.h]: Yeah, I think it is clearer to have the single cases fields isolated into their own structure, instead of a bunch of random fields in the `ViewInfo` structure with eventual comments.

**Comment by dlibenzi** [comment]
> Looks good. Could we eventually unify our `slice` and `as_strided` implementations? The `at::native` version of `slice` eventually routes to `as_strided`, maybe we should follow a similar path at some point.

`slice()` is different from `as_strided()`. The former correctly slices one dimension, while the latter flattens to R1, offsets, and reshape.
It could be composed by a sequence of existing IRs (which was what the first version was doing), but then we end up with bigger IR graphs in which it becomes less clear to follow what a bunch of IRs are doing (ie, implementing an `as_strided()` or its inverse).
